.github/workflows/admin-preview.yml

```diff
@@ -16,46 +16,109 @@ on:
 
-      - name: Install Turbo CLI
-        run: npm install --global turbo@latest
+      - name: Restore node_modules cache
+        id: restore_dependencies
+        uses: actions/cache@v3
+        with:
+          path: ${{ env.CACHED_DEPENDENCY_PATHS }}
+          key: ${{ runner.os }}-modules-${{ steps.compute_lockfile_hash.outputs.key }}
+          restore-keys: |
+            ${{ runner.os }}-modules-
+            ${{ runner.os }}-
 
       - name: Install dependencies
+        if: steps.restore_dependencies.outputs.cache-hit != 'true'
         run: yarn install
+    outputs:
+      dependencies_cache_key: ${{ runner.os }}-modules-${{ steps.compute_lockfile_hash.outputs.key }}
+
+  job_lint:
+    name: Lint
+    needs: job_install_dependencies
+    runs-on: ubuntu-latest
 
-      - name: Build
-        run: turbo build --filter=ui --filter=admin
+    steps:
+      - name: Check out code
+        uses: actions/checkout@v3
+
+      - name: Check dependencies cache
+        uses: actions/cache@v3
+        with:
+          path: ${{ env.CACHED_DEPENDENCY_PATHS }}
+          key: ${{ needs.job_install_dependencies.outputs.dependencies_cache_key }}
 
       - name: Lint
-        run : turbo lint
+        run: yarn lint
 
-  Admin-Deploy-Preview:
-    needs: Admin-Build
+  job_build:
+    name: Build
+    needs: job_install_dependencies
     runs-on: ubuntu-latest
+
     steps:
-      - uses: actions/checkout@v3
-    
+      - name: Check out code
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 2
+
+      - name: Check dependencies cache
+        uses: actions/cache@v3
+        with:
+          path: ${{ env.CACHED_DEPENDENCY_PATHS }}
+          key: ${{ needs.job_install_dependencies.outputs.dependencies_cache_key }}
+
+      - name: Restore bundle
+        id: restore_bundle
+        uses: actions/cache@v3
+        with:
+          path: ${{ env.CACHED_DEPENDENCY_PATHS }}
+          key: ${{ env.BUILD_CACHE_DEPENDENCY_KEYS }}-${{ github.ref_name }}
+
+      - name: Build DEV bundle
+        if: steps.restore_bundle.outputs.cache-hit != 'true'
+        run: yarn build --filter=admin...
+
+  job_deploy_preview:
+    name: Deploy to Preview
+    needs: [job_install_dependencies, job_lint, job_build]
+    runs-on: ubuntu-latest
+    env:
+      VERCEL_ORG_ID: ${{ secrets.VERCEL_ORG_ID }}
+      VERCEL_PROJECT_ID: ${{ secrets.VERCEL_ADMIN_PROJECT_ID }}
+    steps:
+      - name: Check out code
+        uses: actions/checkout@v3
+
       - name: Install Vercel CLI
         run: npm install --global vercel@latest
-      
+
+      - name: Check bundle cache
+        uses: actions/cache@v3
+        with:
+          path: ${{ env.CACHED_DEPENDENCY_PATHS }}
+          key: ${{ env.BUILD_CACHE_DEPENDENCY_KEYS }}-${{ github.ref_name }}
+
+      - name: Build DEV bundle
+        run: yarn build --filter=admin... # 무조건 풀 캐싱 적용 되는 부분임, pull 직전에 turbo run build를 해줘야 종속성이 안 꼬임
```

Comment:
vercel 빌드된 결과물을 github action에서 vercel build 해서 바로 보내는데 여기서 cache를 적용하게 된 이유는 어떤게 있을까요? 그리고 vercel build를 해주는데 그전에 다시 build를 한번 해줘야 하는 이유는 무엇일까요? 주석만 가지고는 이해하기가 어렵네요.

Reply:
저도 이 부분은 이해가 잘 안가는데 `vercel build` 이전에 `turbo build` 가 없으면 ui쪽 모듈을 정상적으로 찾지 못하는 이슈가 있어서 넣어줬습니다.
build를 한 번더 시켜 주더라도 ***풀 캐싱 적용**이 되어서 1초 내외로 build가 완료되기 때문에 성능상 문제는 없지만,
정확한 원인을 파악하게 되면 지우겠습니다!

***풀 캐싱 적용** : 이전 build쪽 job에서 build후 `node_modules/.cache` 폴더가 생기는데 해당 `.cache` 폴더가 있는 `node_modules` 를 캐싱 후 해당 캐싱 파일을 deploy job에서 사용하기 때문에 풀 캐싱 적용이 됩니다.


```Run vercel build --token=***
Vercel CLI 30.0.0
WARNING: You should not upload the `.next` directory.
Installing dependencies...
yarn install v1.22.19
[1/4] Resolving packages...
success Already up-to-date.
Done in 0.60s.
Detected Next.js version: [13](https://github.com/Ssoon-m/frontend/actions/runs/5087924570/jobs/9143787087#step:9:14).4.2
Running "yarn run build"
yarn run v1.22.[19](https://github.com/Ssoon-m/frontend/actions/runs/5087924570/jobs/9143787087#step:9:20)
$ next build
- info Creating an optimized production build...
Failed to compile.

../../packages/ui/dist/index.mjs
Module not found: Can't resolve './chunk-XCQDURPY.mjs'

https://nextjs.org/docs/messages/module-not-found

Import trace for requested module:
./src/app/page.tsx

../../packages/ui/dist/index.mjs
Module not found: Can't resolve './chunk-E[24](https://github.com/Ssoon-m/frontend/actions/runs/5087924570/jobs/9143787087#step:9:25)B3R65.mjs'

https://nextjs.org/docs/messages/module-not-found

Import trace for requested module:
./src/app/page.tsx


> Build failed because of webpack errors
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Error: Command "yarn run build" exited with 1
Error: Process completed with exit code 1.```